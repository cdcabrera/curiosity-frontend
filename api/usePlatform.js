import { useEffect, useMemo } from 'react';
import { useChrome } from '@redhat-cloud-services/frontend-components/useChrome';
import { useUnleashClient } from '@unleash/proxy-client-react';
// import { useShallowCompareEffect } from 'react-use';
// import { featureFlags } from '../../config';

/**
 * Global platform related hooks.
 *
 * @memberof Hooks
 * @module usePlatform
 */

/**
 * @apiMock {DelayResponse} 1000
 * @api {get} /api/featureflags/v0
 * @apiDescription Mock unleash feature flag response.
 *
 * @apiSuccessExample {json} Success-Response:
 *     HTTP/1.1 200 OK
 *     {
 *       "toggles": [
 *         {
 *             "name": "swatch.maintenance",
 *             "enabled": true,
 *             "variant": {
 *                 "name": "disabled",
 *                 "enabled": true
 *             },
 *             "name": "dolor.sit",
 *             "enabled": false,
 *             "variant": {
 *                 "name": "disabled",
 *                 "enabled": false
 *             }
 *         }
 *       ]
 *     }
 *
 * @apiError {Array} errors
 * @apiErrorExample {json} Error-Response:
 *     HTTP/1.1 500 Internal Server Error
 *     {
 *        "errors": []
 *     }
 */
/**
 * Apply a resize observer to an element.
 *
 * @param {string|Array|object} flag
 * @returns {boolean}
 */
const useFeatureFlag = flag => {
  // const woot = useFlag(flag);
  // const test = useFlag(flag);
  // const [updatedFlag, setUpdatedFlag] = useState(false);
  const client = useUnleashClient(); // have to confirm their context updates when flags are updated... hard to test context updates with mocks
  // console.log('>>> client', woot);

  useEffect(() => {
    const testing = client.isEnabled('dolor.sit');
    console.log(testing);
  }, [client]);

  return useMemo(() => {
    if (typeof flag === 'string') {
      return client.isEnabled(flag);
    }
    if (Array.isArray(flag)) {
      return flag.map(value => ({ flag: value, isEnabled: client.isEnabled(value) }));
    }
    if (Object.keys(flag).length) {
      const updatedFlag = { ...flag };
      Object.keys(flag).forEach(value => {
        updatedFlag[value] = client.isEnabled(value);
      });
      return updatedFlag;
    }
  }, [client, flag]);
  /*
  useShallowCompareEffect(() => {
    const result = client.isEnabled(flag);
    console.log('>>>> test', result);
    setUpdatedFlag(result);
  }, [flag]);
  */

  /*
  if (flag) {
    const test = client.isEnabled(flag);
    // const otherTest = client.isEnabled(flag);
    console.log('>>>> test', test);
    return test;
  }
  */

  // return updatedFlag;
};

// const useFeatureFlags = ({ useDispatch: useAliasDispatch = useDispatch, useFlagClient: useAliasFlagClient = useUnleashClient } = {}) => {
// const dispatch = useDispatch();
// }

const platformHooks = {
  useChrome,
  useFeatureFlag
};

export { platformHooks as default, platformHooks, useChrome, useFeatureFlag };
