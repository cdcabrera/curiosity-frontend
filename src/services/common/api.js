import axios, { CancelToken } from 'axios';
import LruCache from 'lru-cache';
import { serviceHelpers } from './helpers';

/**
 * Cache Axios service call cancel tokens.
 *
 * @private
 * @type {object}
 */
const cancelTokens = {};

/**
 * Cache Axios service call responses.
 *
 * @type {object}
 */
const responseCache = new LruCache({
  maxAge: Number.parseInt(process.env.REACT_APP_AJAX_CACHE, 10) || 30000,
  max: 100,
  updateAgeOnGet: true
});

/**
 * Set Axios configuration, which includes response schema validation and caching.
 * Call platform "getUser" auth method, and apply service config. Service configuration
 * includes the ability to cancel all and specific calls, cache and normalize a response
 * based on both a provided schema and a successful API response. The cache will refresh
 * its timeout on continuous calls. To reset it a user will either need to refresh the
 * page or wait the "maxAge".
 *
 * @param {object} config
 * @param {boolean} config.authenticate
 * @param {string} config.url
 * @param {object} config.params
 * @param {boolean} config.cache
 * @param {boolean} config.cancel
 * @param {string} config.cancelId
 * @param {Array} config.schema
 * @param {Array} config.transform
 * @returns {Promise<*>}
 */
const serviceCall = async (config = {}) => {
  const updatedConfig = {
    ...config,
    cache: undefined,
    cacheResponse: config.cache
  };

  const responseTransformers = [];
  const cancelledMessage = 'cancelled request';
  const axiosInstance = axios.create();
  const sortedParams =
    (updatedConfig.params && Object.entries(updatedConfig.params).sort(([a], [b]) => a.localeCompare(b))) || [];
  const cacheId = `${updatedConfig.url || ''}-${JSON.stringify(sortedParams)}`;

  if (updatedConfig.cancel === true) {
    const cancelTokensId = `${updatedConfig.cancelId || ''}-${updatedConfig.url}`;

    if (cancelTokens[cancelTokensId]) {
      cancelTokens[cancelTokensId].cancel(cancelledMessage);
    }

    cancelTokens[cancelTokensId] = CancelToken.source();
    updatedConfig.cancelToken = cancelTokens[cancelTokensId].token;

    delete updatedConfig.cancel;
  }

  if (updatedConfig.cacheResponse === true) {
    const cachedResponse = responseCache.get(cacheId);

    if (cachedResponse) {
      updatedConfig.adapter = adapterConfig =>
        Promise.resolve({
          ...cachedResponse,
          status: 304,
          statusText: 'Not Modified',
          config: adapterConfig
        });

      return axiosInstance(updatedConfig);
    }
  }

  if (updatedConfig.schema) {
    responseTransformers.push(updatedConfig.schema);
  }

  if (updatedConfig.transform) {
    responseTransformers.push(updatedConfig.transform);
  }

  responseTransformers.forEach(([successTransform, errorTransform]) => {
    const transformers = [undefined, response => Promise.reject(response)];

    if (successTransform) {
      transformers[0] = response => {
        const updatedResponse = { ...response };
        const { data, error: normalizeError } = serviceHelpers.passDataToCallback(
          updatedResponse.data,
          successTransform
        );

        if (!normalizeError) {
          updatedResponse.data = data;
        }

        return updatedResponse;
      };
    }

    if (errorTransform) {
      transformers[1] = response => {
        const updatedResponse = { ...response };

        if (updatedResponse?.message === cancelledMessage) {
          return Promise.reject(updatedResponse);
        }

        const { data, error: normalizeError } = serviceHelpers.passDataToCallback(
          updatedResponse?.response?.data,
          errorTransform
        );

        if (!normalizeError) {
          updatedResponse.response = { ...updatedResponse.response, data };
        }

        return Promise.reject(updatedResponse);
      };
    }

    axiosInstance.interceptors.response.use(...transformers);
  });

  if (updatedConfig.cacheResponse === true) {
    axiosInstance.interceptors.response.use(
      response => {
        const updatedResponse = { ...response };
        responseCache.set(cacheId, updatedResponse);
        return updatedResponse;
      },
      response => Promise.reject(response)
    );
  }

  if (typeof updatedConfig.url === 'function') {
    const emulateCallback = updatedConfig.url;
    updatedConfig.url = '/';

    let message = `success, emulated`;
    let emulatedResponse;
    let isSuccess = true;

    try {
      emulatedResponse = await emulateCallback();
    } catch (e) {
      isSuccess = false;
      message = e.message;
    }

    if (isSuccess) {
      updatedConfig.adapter = adapterConfig =>
        Promise.resolve({
          data: emulatedResponse,
          status: 200,
          statusText: message,
          config: adapterConfig
        });
    } else {
      updatedConfig.adapter = adapterConfig =>
        Promise.reject({ // eslint-disable-line
          ...new Error(message),
          message,
          status: 418,
          config: adapterConfig
        });
    }
  }

  console.log('API >>>>>>>', updatedConfig);

  return axiosInstance(updatedConfig);
};

const api = { serviceCall };

export { api as default, api, serviceCall };
