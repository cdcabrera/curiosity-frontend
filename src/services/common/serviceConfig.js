import axios, { CancelToken } from 'axios';
import LruCache from 'lru-cache';
import { serviceHelpers } from './helpers';

/**
 * Set Axios XHR default timeout.
 */
const globalXhrTimeout = Number.parseInt(process.env.REACT_APP_AJAX_TIMEOUT, 10) || 60000;

/**
 * Cache Axios service call cancel tokens.
 *
 * @private
 * @type {object}
 */
const globalCancelTokens = {};

/**
 * Cache Axios service call responses.
 *
 * @type {object}
 */
const globalResponseCache = new LruCache({
  maxAge: Number.parseInt(process.env.REACT_APP_AJAX_CACHE, 10) || 30000,
  max: 100,
  updateAgeOnGet: true
});

/**
 * Set modified Axios configuration and instance. Additions include
 * - schema validation
 * - response transformations
 * - caching plain responses, caching of validated and transformed response
 * - automatic cancelling if the same call is fired in succession
 *
 * @param {object} config
 * @param {boolean} config.cache
 * @param {boolean} config.cancel
 * @param {string} config.cancelId
 * @param {object} config.headers
 * @param {object} config.params
 * @param {Array} config.schema
 * @param {number} config.timeout
 * @param {Array} config.transform
 * @param {string|Function} config.url Provide an API path, or a function that returns a promise/promise like function, or emulated API response.
 */
/*
class AxiosConfigPROBSWITHASYNC {
  axiosInstance = axios.create();

  config = {};

  cancelledMessage = 'cancelled request';

  cancelTokens = globalCancelTokens;

  responseCache = globalResponseCache;

  xhrTimeout = globalXhrTimeout;

  constructor(config) {
    this.config = {
      timeout: this.xhrTimeout,
      ...config,
      cache: undefined,
      cacheResponse: config.cache
    };
  }

  serviceCall() {
    const { axiosInstance, cacheId, config, responseCache } = this;
    const updatedConfig = { ...config };

    this.setCacheId();
    this.setCancel();

    if (updatedConfig.cacheResponse === true) {
      const cachedResponse = responseCache.get(cacheId);

      if (cachedResponse) {
        updatedConfig.adapter = adapterConfig =>
          Promise.resolve({
            ...cachedResponse,
            status: 304,
            statusText: 'Not Modified',
            config: adapterConfig
          });

        return axiosInstance(updatedConfig);
      }
    }

    this.setTransformations();

    if (updatedConfig.cacheResponse === true) {
      axiosInstance.interceptors.response.use(
        response => {
          const updatedResponse = { ...response };
          responseCache.set(cacheId, updatedResponse);
          return updatedResponse;
        },
        response => Promise.reject(response)
      );
    }

    (async () => this.setEmulatedService())();
    // this.setEmulatedService();

    console.log('CLASS AXIOS CONFIG >>>>>>>', updatedConfig);

    return (
      (updatedConfig.url && axiosInstance(updatedConfig)) || Promise.reject(new Error('Axios config.url missing.'))
    );
  }

  setCacheId() {
    const { config } = this;
    const sortedParams = (config.params && Object.entries(config.params).sort(([a], [b]) => a.localeCompare(b))) || [];

    this.cacheId = `${config.url || ''}-${JSON.stringify(sortedParams)}`;
  }

  setCancel() {
    const { config, cancelledMessage, cancelTokens } = this;
    const updatedConfig = { ...config };

    if (updatedConfig.cancel === true) {
      const cancelTokensId = `${updatedConfig.cancelId || ''}-${updatedConfig.url}`;

      if (cancelTokens[cancelTokensId]) {
        cancelTokens[cancelTokensId].cancel(cancelledMessage);
      }

      cancelTokens[cancelTokensId] = CancelToken.source();
      updatedConfig.cancelToken = cancelTokens[cancelTokensId].token;

      delete updatedConfig.cancel;
    }

    this.config = updatedConfig;
  }

  setTransformations() {

  }

  async setEmulatedService() {

  }
}
*/

/**
 * Generate a cache ID.
 *
 * @param {object} config
 * @param {object} config.params
 * @param {string|Function} config.url
 * @returns {string}
 */
const setCacheId = (config = {}) => {
  const sortedParams = (config.params && Object.entries(config.params).sort(([a], [b]) => a.localeCompare(b))) || [];
  return `${config.url || ''}-${JSON.stringify(sortedParams)}`;
};

/**
 * Generate a cancel token.
 *
 * @param {object} config
 * @param {boolean} config.cancel
 * @param {string} config.cancelId
 * @param {string|Function} config.url
 * @param {object} options
 * @param {string} options.cancelledMessage
 * @param {object} options.cancelTokens
 * @returns {object}
 */
const setCancel = (config = {}, { cancelledMessage = '', cancelTokens } = {}) => {
  const updatedConfig = { ...config };
  const updatedCancelTokens = cancelTokens;

  if (updatedConfig.cancel === true) {
    const cancelTokensId = `${updatedConfig.cancelId || ''}-${updatedConfig.url}`;

    if (updatedCancelTokens[cancelTokensId]) {
      updatedCancelTokens[cancelTokensId].cancel(cancelledMessage);
    }

    updatedCancelTokens[cancelTokensId] = CancelToken.source();
    updatedConfig.cancelToken = updatedCancelTokens[cancelTokensId].token;

    delete updatedConfig.cancel;
  }

  return updatedConfig;
};

/**
 * Confirm transformation config, format, return an array of transformations (schema checks, response transforms).
 *
 * @param {object} config
 * @param {Array} config.schema
 * @param {Array} config.transform
 * @param {object} options
 * @param {string} options.cancelledMessage
 * @returns {Array}
 */
const setTransformations = (config = {}, { cancelledMessage = '' } = {}) => {
  const responseTransformers = [];

  if (config.schema) {
    responseTransformers.push(config.schema);
  }

  if (config.transform) {
    responseTransformers.push(config.transform);
  }

  return responseTransformers.map(([successTransform, errorTransform]) => {
    const transformers = [undefined, response => Promise.reject(response)];

    if (successTransform) {
      transformers[0] = response => {
        const updatedResponse = { ...response };
        const { data, error: normalizeError } = serviceHelpers.passDataToCallback(
          updatedResponse.data,
          successTransform
        );

        if (!normalizeError) {
          updatedResponse.data = data;
        }

        return updatedResponse;
      };
    }

    if (errorTransform) {
      transformers[1] = response => {
        const updatedResponse = { ...response };

        if (updatedResponse?.message === cancelledMessage) {
          return Promise.reject(updatedResponse);
        }

        const { data, error: normalizeError } = serviceHelpers.passDataToCallback(
          updatedResponse?.response?.data,
          errorTransform
        );

        if (!normalizeError) {
          updatedResponse.response = { ...updatedResponse.response, data };
        }

        return Promise.reject(updatedResponse);
      };
    }

    return transformers;
    // this.axiosInstance.interceptors.response.use(...transformers);
  });
};

/**
 * Return an emulated service response with a function. Applies a consistent way of returning data in scenarios where
 * global functions are provided.
 *
 * @param {object} config
 * @param {string|Function} config.url
 * @returns {Promise}
 */
const setEmulatedService = async config => {
  const updatedConfig = { ...config };

  if (typeof updatedConfig.url === 'function') {
    const emulateCallback = updatedConfig.url;
    updatedConfig.url = '/';

    let message = `success, emulated`;
    let emulatedResponse;
    let isSuccess = true;

    try {
      emulatedResponse = await emulateCallback();
    } catch (e) {
      isSuccess = false;
      message = e.message;
    }

    if (isSuccess) {
      updatedConfig.adapter = adapterConfig =>
        Promise.resolve({
          data: emulatedResponse,
          status: 200,
          statusText: message,
          config: adapterConfig
        });
    } else {
      updatedConfig.adapter = adapterConfig =>
        Promise.reject({ // eslint-disable-line
          ...new Error(message),
          message,
          status: 418,
          config: adapterConfig
        });
    }
  }

  return updatedConfig;
};

/**
 * Set Axios configuration.
 *
 * @param {object} config
 * @param {object} config.cache
 * @param {boolean} config.cancel
 * @param {string} config.cancelId
 * @param {object} config.params
 * @param {Array} config.schema
 * @param {Array} config.transform
 * @param {string|Function} config.url
 * @param {object} options
 * @param {string} options.cancelledMessage
 * @param {object} options.cancelTokens
 * @param {object} options.responseCache
 * @param {number} options.xhrTimeout
 * @returns {Promise}
 */
const axiosConfig = async (
  config = {},
  {
    cancelledMessage = 'cancelled request',
    cancelTokens = globalCancelTokens,
    responseCache = globalResponseCache,
    xhrTimeout = globalXhrTimeout
  } = {}
) => {
  const axiosInstance = axios.create();
  let updatedConfig = {
    timeout: xhrTimeout,
    ...config,
    cache: undefined,
    cacheResponse: config.cache
  };

  const cacheId = setCacheId(updatedConfig);

  updatedConfig = setCancel(updatedConfig, { cancelledMessage, cancelTokens });

  if (updatedConfig.cacheResponse === true) {
    const cachedResponse = responseCache.get(cacheId);

    if (cachedResponse) {
      updatedConfig.adapter = adapterConfig =>
        Promise.resolve({
          ...cachedResponse,
          status: 304,
          statusText: 'Not Modified',
          config: adapterConfig
        });

      return axiosInstance(updatedConfig);
    }
  }

  setTransformations(updatedConfig, { cancelledMessage }).forEach(transformers =>
    axiosInstance?.interceptors?.response?.use(...transformers)
  );

  if (updatedConfig.cacheResponse === true) {
    axiosInstance?.interceptors?.response?.use(
      response => {
        const updatedResponse = { ...response };
        responseCache.set(cacheId, updatedResponse);
        return updatedResponse;
      },
      response => Promise.reject(response)
    );
  }

  updatedConfig = await setEmulatedService(updatedConfig);

  console.log('REWRITE AXIOS CONFIG >>>>>>>', updatedConfig);

  // return (updatedConfig.url && axiosInstance(updatedConfig)) || Promise.reject(new Error('Axios config.url missing.'));
  return axiosInstance(updatedConfig);
};

const serviceConfig = { axiosConfig, globalXhrTimeout, globalCancelTokens, globalResponseCache };

export {
  serviceConfig as default,
  serviceConfig,
  axiosConfig,
  globalXhrTimeout,
  globalCancelTokens,
  globalResponseCache
};
