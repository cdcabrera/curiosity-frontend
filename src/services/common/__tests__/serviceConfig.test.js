import axios from 'axios';
import { serviceConfig, axiosConfig } from '../serviceConfig';
import { serviceCall } from '../api';

describe('ServiceConfig', () => {
  it('should have specific functions', () => {
    expect(serviceConfig).toMatchSnapshot('serviceConfig');
  });

  it('should do things consistently', () => {
    expect(serviceCall({})).toMatchSnapshot('001, original');
    // expect(new AxiosConfig({}).serviceCall()).toMatchSnapshot('001, class');
    expect(axiosConfig({})).toMatchSnapshot('001, funcs');
  });

  it('should pass a configuration to Axios', () => {
    // expect(serviceCall({})).toMatchSnapshot('001, original');
    // expect(new AxiosConfig({}).serviceCall()).toMatchSnapshot('001, class');
    const mockAxiosInstance = jest.fn();
    // const spy = jest.spyOn(axios, 'create').mockImplementation(config => ({ config }));
    // const spy = jest.spyOn(axios, 'create').mockImplementation(() => config => ({ config }));
    // const spy = jest.spyOn(axios, 'create').mockImplementation(() => config => jest.fn(config));
    const spy = jest.spyOn(axios, 'create').mockImplementation(() => {
      console.log('TEST >>> 001');
      return config => {
        console.log('TEST >>> 002', config);
        return {
          ...config
        };
      };
    });

    axiosConfig({ url: '/', cache: true, cancel: true, schema: [() => {}], transform: [() => {}] });
    expect(mockAxiosInstance.mock).toMatchSnapshot('002, function');
    // expect(spy.mock.calls).toMatchSnapshot('002, function');

    // mockAxiosInstance.mockClear();

    // new AxiosConfig({ url: '/', cache: true, cancel: true, schema: [() => {}], transform: [() => {}] }).serviceCall();
    // expect(mockAxiosInstance.mock.calls).toMatchSnapshot('002, class');

    // mockAxiosInstance.mockClear();
    spy.mockClear();
  });
});
