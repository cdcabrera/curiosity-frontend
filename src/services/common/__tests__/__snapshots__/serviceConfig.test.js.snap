// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceConfig should do things consistently: 001, funcs 1`] = `Promise {}`;

exports[`ServiceConfig should do things consistently: 001, original 1`] = `Promise {}`;

exports[`ServiceConfig should have specific functions: serviceConfig 1`] = `
Object {
  "axiosConfig": [Function],
  "globalCancelTokens": Object {},
  "globalResponseCache": LRUCache {
    Symbol(max): 100,
    Symbol(lengthCalculator): [Function],
    Symbol(allowStale): false,
    Symbol(maxAge): 30000,
    Symbol(dispose): undefined,
    Symbol(noDisposeOnSet): false,
    Symbol(updateAgeOnGet): true,
    Symbol(cache): Map {},
    Symbol(lruList): Yallist {
      "head": null,
      "length": 0,
      "tail": null,
    },
    Symbol(length): 0,
  },
  "globalXhrTimeout": 60000,
}
`;

exports[`ServiceConfig should pass a configuration to Axios: 002, function 1`] = `
Array [
  Array [],
]
`;
